README.md
# NAME

Net::Kubernetes - An object oriented interface to the REST API's provided by kubernetes

[![Build Status](https://travis-ci.org/perljedi/net-kubernetes.png?branch=master)](https://travis-ci.org/perljedi/net-kubernetes)

# VERSION

version 1.03

# SYNOPSIS

    my $kube = Net::Kubernetes->new(url=>'http://127.0.0.1:8080', username=>'dave', password=>'davespassword');
    my $pod_list = $kube->list_pods();

    my $nginx_pod = $kube->create_from_file('kubernetes/examples/pod.yaml');

    my $ns = $kube->get_namespace('default');

    my $services = $ns->list_services;

    my $pod = $ns->get_pod('my-pod');

    $pod->delete;

    my $other_pod = $ns->create_from_file('./my-pod.yaml');

# METHODS

## new - Create a new $kube object

All parameters are optional and have some basic default values (where appropriate).

- url \['http://localhost:8080'\]

    The base url for the kubernetes. This should include the protocal (http or https) but not "/api/v1beta3" (see base\_path).

- base\_path \['/api/v1beta3'\]

    The entry point for api calls, this may be used to set the api version with which to interact.

- username

    Username to use with basic authentication. If either username or password are not provided, basic authentication will not
    be used.

- password

    Password to use with basic authentication. If either username or password are not provided, basic authentication will not
    be used.

- token

    An authentication token to be used to access the apiserver.  This may be provided as a plain string, a path to a file
    from which to read the token (like /var/run/secrets/kubernetes.io/serviceaccount/token from within a pod), or a reference
    to a file handle (from which to read the token).

- ssl\_cert\_file, ssl\_key\_file, ssl\_ca\_file

    This there options passed into new will cause Net::Kubernetes in inlcude SSL client certs to requests to the kuberernetes
    API server for authentication.  There are basically just a passthrough to the underlying LWP::UserAgent used to handle the 
    api requests.

## get\_namespace("myNamespace");

This method returns a "Namespace" object on which many methods can be called implicitly
limited to the specified namespace.

## get\_pod('my-pod-name')

Delegates automatically to [Net::Kubernetes::Namespace](https://metacpan.org/pod/Net::Kubernetes::Namespace) via $self->get\_namespace('default')

## get\_repllcation\_controller('my-rc-name') (aliased as $ns->get\_rc('my-rc-name'))

Delegates automatically to [Net::Kubernetes::Namespace](https://metacpan.org/pod/Net::Kubernetes::Namespace) via $self->get\_namespace('default')

## get\_service('my-servce-name')

Delegates automatically to [Net::Kubernetes::Namespace](https://metacpan.org/pod/Net::Kubernetes::Namespace) via $self->get\_namespace('default')

## get\_secret('my-secret-name')

Delegates automatically to [Net::Kubernetes::Namespace](https://metacpan.org/pod/Net::Kubernetes::Namespace) via $self->get\_namespace('default')

## list\_nodes(\[label=>{label=>value}\], \[fields=>{field=>value}\])

returns a list of [Net::Kubernetes::Resource::Node](https://metacpan.org/pod/Net::Kubernetes::Resource::Node)s

## list\_service\_accounts(\[label=>{label=>value}\], \[fields=>{field=>value}\])

returns a list of [Net::Kubernetes::Resource::Service](https://metacpan.org/pod/Net::Kubernetes::Resource::Service)s

<div>
    <h2>Build Status</h2>

    <img src="https://travis-ci.org/perljedi/net-kubernetes.svg?branch=release-0.21" />
</div>

# AUTHOR

Dave Mueller &lt;dave@perljedi.com>

# COPYRIGHT AND LICENSE

This software is Copyright (c) 2015 by Dave Mueller.

This is free software, licensed under:

    The MIT (X11) License

# SEE ALSO

Please see those modules/websites for more information related to this module.

- [Net::Kubernetes::Namespace](https://metacpan.org/pod/Net::Kubernetes::Namespace)
- [Net::Kubernetes::Resource](https://metacpan.org/pod/Net::Kubernetes::Resource)

# CONSUMES

- [Net::Kubernetes::Role::APIAccess](https://metacpan.org/pod/Net::Kubernetes::Role::APIAccess)
- [Net::Kubernetes::Role::ResourceFactory](https://metacpan.org/pod/Net::Kubernetes::Role::ResourceFactory)
- [Net::Kubernetes::Role::ResourceFetcher](https://metacpan.org/pod/Net::Kubernetes::Role::ResourceFetcher)
- [Net::Kubernetes::Role::ResourceLister](https://metacpan.org/pod/Net::Kubernetes::Role::ResourceLister)

# CONTRIBUTORS

- Christopher Pruden &lt;cdpruden@liquidweb.com>
- Dave &lt;dave@perljedi.com>
- Dave Mueller &lt;dmueller@liquidweb.com>
- Kevin Johnson &lt;kcavemanj@gmail.com>


Changes
Revision history for {{$dist->name}}

{{$NEXT}}

1.03 Tue Jul 12 2016

  - Bug fix for boolean values passed to kubernetes via json.

1.02 Wed May 11 2016
 
  - Update dependency on Test::Mock::Wrapper to avoid bug in perl 5.24

1.01 Thu Apr 28 2016

  - Resource factory now ensures 'kind' is populated

1.0 Thu Apr 28 2016

  - Resource->as_hashref now returns a clone of the data.

0.23 Mon Apr 18 2016

  - Remove "resource version" in as hashref to prevent update resource bug.

0.22 Mon Feb 22 2016

  - Add support for SSL client certificate based
    authhenitcation.

  - Minor documentation tweaks

0.21 Mon Sep 14 2015

 - Require Test::Mock::Wrapper v0.17 or better for tests
   (https://rt.cpan.org/Public/Bug/Display.html?id=107091)

0.20 Mon Sep 14 2015

 - Update serialized version of endpoitn resource
 
 - Update the UA string in http headers

0.19 Mon Aug 10 2015

 - Correct cirtical URL typo in list_endpoints

0.18 Fri Aug 07 2015

 - Add ServiceAccount resource, and list method to fetch it

 - Add get_node method

 - Add get_pods method to node object

 - Refactor some tests to reduce code duplication

0.17 Fri Jul 31 2015

 - Remove errant debuging statement

 - Add "scale" method to Replication Controller

0.16 Fri Jul 31 2015

 - Bug fix in get_pods to correctly AND together labels in the label selector

 - Typo fix in POD

0.15 Thu Jul 30 2015

 - Update for compatibility with kubernetes v1 release

0.14 Sat Jul 18 2015

 - Added "logs" access to pod resource object.

 - Added "Node" resource type and "list_nodes" method.

 - Added "Endpoint" resource type and "list_endpoints" method.

 - Fixed broken tests for win32 systems.

0.13 Tue Jul 14 2015

 - Housekeeping - Make sure contributors are credited in pod and readme correctly

0.12 Tue Jul 14 2015

 - New resource type "Event"

 - New "list_events" methods

 - Updated documentation on creating secrets

0.11 Tue Jun 30 2015

 - Lots of documentation additions and enhacement

 - Added several sub types of the Net::Kubernetes::Exception
   to allow for cleaner/easier exception handling.

 - Update $kube to also delegate create and create_from file
   to the default namespace.

 - Added "build_secret" method to make creating secrets easier

0.10 Sat Jun 27 2015

 - Fixed circular dependency between Net::Kubernetes::Resource::Role::HasPods
   and Net::Kubernetes::Role::ResourceFactory which caused bugs in get_pods

 - Added automatic delegation from the top level $kube object
   to default namespace for get_pod, get_secret, get_rc and
   get_service

0.09 Fri Jun 26 2015

 - Standardize url construction (which also fixes a few bugs)

 - Minor resource role refactor for clarity

 - Added lots of unit tests for resource objects

 - Adding travis-ci build status image to POD

 - Added "get_pods" method to Service Resource

 - Refactor docs to a more sane layout.

0.08 Thu Jun 25 2015

 - Added new method to resource controller to fetch its pods

 - Added test cases, refactored some tests to re-use code

 - Removed references to Data::Dumper that were only sued
   for debugging.

0.06 Tue Jun 23 2015

 - Added a dzil plugin to handle making meta.yaml

 - Added an abstract in the main module to clear
   cpants kwalitee check

 - Switched to use Cpanel::JSON::XS for stability

 - Swtich to YAML::XS which appears to have better
   handling for numbers read from a yaml file.

 - Temporary fix for conversion problems
   between YAML and JSON in resource creator.

0.05 Sun Jun 21 2015

 - Apparently cpan thinks 0.04.1 is less than 0.04

0.04.1 Sun Jun 21 2015

 - Fix bug causing resource and namespace urls to get a double /

0.04 Sat Jun 20 2015

 - Fix bug reading token in perl 5.18 reported by cpan testers

 - Fix a bug in reading json from disk in create_from_file

0.03 Fri Jun 19 2015

 - New authentication option: "Token"
      provide a token to use when authentication to the kubernetes apiserver
      may be provided as a string, path to a file, or a ref to a file handle

 - Resource method "update"
      Any reosurce object may be modified (using the provided attributes).
      after making changes, call $resource->update; to send those changes
      to the api server to store.


0.02 Sun Jun 07 2015

 - Fixed typo in documentation
 - Removed debug code from delete method
 - Added refresh method for statefull resources.

0.01 Sat Jun 06 2015


lib/Net/Kubernetes.pm
package Net::Kubernetes;
# ABSTRACT: An object oriented interface to the REST API's provided by kubernetes

use Moose;
require Net::Kubernetes::Namespace;
require LWP::UserAgent;
require HTTP::Request;
require URI;;
require Throwable::Error;
use MIME::Base64;
require Net::Kubernetes::Exception;

=head1 SYNOPSIS

  my $kube = Net::Kubernetes->new(url=>'http://127.0.0.1:8080', username=>'dave', password=>'davespassword');
  my $pod_list = $kube->list_pods();

  my $nginx_pod = $kube->create_from_file('kubernetes/examples/pod.yaml');

  my $ns = $kube->get_namespace('default');

  my $services = $ns->list_services;

  my $pod = $ns->get_pod('my-pod');

  $pod->delete;

  my $other_pod = $ns->create_from_file('./my-pod.yaml');

=begin html

<h2>Build Status</h2>

<img src="https://travis-ci.org/perljedi/net-kubernetes.svg?branch=release-0.21" />

=end html

=cut

with 'Net::Kubernetes::Role::APIAccess';
with 'Net::Kubernetes::Role::ResourceLister';
with 'Net::Kubernetes::Role::ResourceFetcher';

=method new - Create a new $kube object

All parameters are optional and have some basic default values (where appropriate).

=over 1

=item url ['http://localhost:8080']

The base url for the kubernetes. This should include the protocal (http or https) but not "/api/v1beta3" (see base_path).

=item base_path ['/api/v1beta3']

The entry point for api calls, this may be used to set the api version with which to interact.

=item username

Username to use with basic authentication. If either username or password are not provided, basic authentication will not
be used.

=item password

Password to use with basic authentication. If either username or password are not provided, basic authentication will not
be used.

=item token

An authentication token to be used to access the apiserver.  This may be provided as a plain string, a path to a file
from which to read the token (like /var/run/secrets/kubernetes.io/serviceaccount/token from within a pod), or a reference
to a file handle (from which to read the token).

=item ssl_cert_file, ssl_key_file, ssl_ca_file

This there options passed into new will cause Net::Kubernetes in inlcude SSL client certs to requests to the kuberernetes
API server for authentication.  There are basically just a passthrough to the underlying LWP::UserAgent used to handle the 
api requests.

=back

=method get_namespace("myNamespace");

This method returns a "Namespace" object on which many methods can be called implicitly
limited to the specified namespace.

=method get_pod('my-pod-name')

Delegates automatically to L<Net::Kubernetes::Namespace> via $self->get_namespace('default')

=method get_repllcation_controller('my-rc-name') (aliased as $ns->get_rc('my-rc-name'))

Delegates automatically to L<Net::Kubernetes::Namespace> via $self->get_namespace('default')

=method get_service('my-servce-name')

Delegates automatically to L<Net::Kubernetes::Namespace> via $self->get_namespace('default')

=method get_secret('my-secret-name')

Delegates automatically to L<Net::Kubernetes::Namespace> via $self->get_namespace('default')

=cut

has 'default_namespace' => (
	is         => 'rw',
	isa        => 'Net::Kubernetes::Namespace',
	required   => 0,
	lazy       => 1,
	handles    => [qw(get_pod get_rc get_replication_controller get_secret get_service create create_from_file build_secret)],
	builder    => '_get_default_namespace',
);

sub get_namespace {
	my($self, $namespace) = @_;
	if (! defined $namespace || ! length $namespace) {
		Throwable::Error->throw(message=>'$namespace cannot be null');
	}
	my $res = $self->ua->request($self->create_request(GET => $self->path.'/namespaces/'.$namespace));
	if ($res->is_success) {
		my $ns = $self->json->decode($res->content);
		my(%create_args) = (url => $self->url, base_path=>$ns->{metadata}{selfLink}, api_version=>$self->api_version, namespace=> $namespace, _namespace_data=>$ns);
		$create_args{username} = $self->username if(defined $self->username);
		$create_args{password} = $self->password if(defined $self->password);
		$create_args{ssl_cert_file} = $self->ssl_cert_file if(defined $self->ssl_cert_file);
		$create_args{ssl_key_file} = $self->ssl_key_file if(defined $self->ssl_key_file);
		$create_args{ssl_ca_file} = $self->ssl_ca_file if(defined $self->ssl_ca_file);
		return Net::Kubernetes::Namespace->new(%create_args);
	}else{
		Net::Kubernetes::Exception->throw(code=>$res->code, message=>"Error getting namespace $namespace:\n".$res->message);
	}
}

=method list_nodes([label=>{label=>value}], [fields=>{field=>value}])

returns a list of L<Net::Kubernetes::Resource::Node>s

=cut

sub list_nodes {
	my $self = shift;
	my(%options);
	if (ref($_[0])) {
		%options = %{ $_[0] };
	}else{
		%options = @_;
	}

	my $uri = URI->new($self->path.'/nodes');
	my(%form) = ();
	$form{labelSelector}=$self->_build_selector_from_hash($options{labels}) if (exists $options{labels});
	$form{fieldSelector}=$self->_build_selector_from_hash($options{fields}) if (exists $options{fields});
	$uri->query_form(%form);

	my $res = $self->ua->request($self->create_request(GET => $uri));
	if ($res->is_success) {
		my $node_list = $self->json->decode($res->content);
		my(@nodes)=();
		foreach my $node (@{ $node_list->{items}}){
			$node->{apiVersion} = $node_list->{apiVersion};
			push @nodes, $self->create_resource_object($node, 'Node');
		}
		return wantarray ? @nodes : \@nodes;
	}else{
		Net::Kubernetes::Exception->throw(code=>$res->code, message=>$res->message);
	}
}

sub get_node {
	my($self, $name) = @_;
	Net::Kubernetes::Exception->throw(message=>"Missing required parameter 'name'") if(! defined $name || ! length $name);
	return $self->get_resource_by_name($name, 'nodes');
}

=method list_service_accounts([label=>{label=>value}], [fields=>{field=>value}])

returns a list of L<Net::Kubernetes::Resource::Service>s

=cut

sub list_service_accounts {
	my $self = shift;
	my(%options);
	if (ref($_[0])) {
		%options = %{ $_[0] };
	}else{
		%options = @_;
	}

	my $uri = URI->new($self->path.'/serviceaccounts');
	my(%form) = ();
	$form{labelSelector}=$self->_build_selector_from_hash($options{labels}) if (exists $options{labels});
	$form{fieldSelector}=$self->_build_selector_from_hash($options{fields}) if (exists $options{fields});
	$uri->query_form(%form);

	my $res = $self->ua->request($self->create_request(GET => $uri));
	if ($res->is_success) {
		my $sa_list = $self->json->decode($res->content);
		my(@saccs)=();
		foreach my $sacc (@{ $sa_list->{items}}){
			$sacc->{apiVersion} = $sa_list->{apiVersion};
			push @saccs, $self->create_resource_object($sacc, 'ServiceAccount');
		}
		return wantarray ? @saccs : \@saccs;
	}else{
		Net::Kubernetes::Exception->throw(code=>$res->code, message=>$res->message);
	}
}


sub _get_default_namespace {
	my($self) = @_;
	return $self->get_namespace('default');
}

# SEEALSO: Net::Kubernetes::Namespace, Net::Kubernetes::Resource

return 42;
